packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/BsonEncoder.kt

```diff
@@ -0,0 +1,292 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@file:Suppress("invisible_reference", "invisible_member")
+
+package io.realm.kotlin.mongodb.internal
+
+import io.realm.kotlin.internal.ObjectIdImpl
+import io.realm.kotlin.internal.ObjectIdSerializer
+import io.realm.kotlin.internal.RealmInstantSerializer
+import io.realm.kotlin.internal.RealmUUIDSerializer
+import io.realm.kotlin.internal.toMillis
+import io.realm.kotlin.internal.toRealmInstant
+import io.realm.kotlin.types.MutableRealmInt
+import io.realm.kotlin.types.ObjectId
+import io.realm.kotlin.types.RealmInstant
+import io.realm.kotlin.types.RealmUUID
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.modules.plus
+import kotlinx.serialization.serializer
+import org.mongodb.kbson.BsonArray
+import org.mongodb.kbson.BsonBinary
+import org.mongodb.kbson.BsonBinarySubType
+import org.mongodb.kbson.BsonBoolean
+import org.mongodb.kbson.BsonDBPointer
+import org.mongodb.kbson.BsonDateTime
+import org.mongodb.kbson.BsonDecimal128
+import org.mongodb.kbson.BsonDocument
+import org.mongodb.kbson.BsonDouble
+import org.mongodb.kbson.BsonElement
+import org.mongodb.kbson.BsonInt32
+import org.mongodb.kbson.BsonInt64
+import org.mongodb.kbson.BsonInvalidOperationException
+import org.mongodb.kbson.BsonJavaScript
+import org.mongodb.kbson.BsonJavaScriptWithScope
+import org.mongodb.kbson.BsonMaxKey
+import org.mongodb.kbson.BsonMinKey
+import org.mongodb.kbson.BsonNull
+import org.mongodb.kbson.BsonNumber
+import org.mongodb.kbson.BsonObjectId
+import org.mongodb.kbson.BsonRegularExpression
+import org.mongodb.kbson.BsonString
+import org.mongodb.kbson.BsonSymbol
+import org.mongodb.kbson.BsonTimestamp
+import org.mongodb.kbson.BsonType
+import org.mongodb.kbson.BsonUndefined
+import org.mongodb.kbson.BsonValue
+
+/**
+ * Bson encoder based on the `Json` encoder from the `KSerializer`. To avoid using any
+ * experimental `KSerializer` APIs and to maximize compatibility it only supports a predefined set
+ * of types:
+ * - Primitives, Realm, Bson, Lists and Map types for encoding.
+ * - Primitives, Realm and Bson types for decoding.
+ */
+public object BsonEncoder {
+    // We need to define these serializers as contextual because they are not available in the
+    // serializer module.
+    private val realmSerializersModule = SerializersModule {
+        contextual(ObjectId::class, ObjectIdSerializer)
+        contextual(RealmUUID::class, RealmUUIDSerializer)
+        contextual(RealmInstant::class, RealmInstantSerializer)
+    }
+
+    public val serializersModule: SerializersModule =
+        Json.serializersModule.plus(realmSerializersModule)
+
+    /**
+     * Encodes a given value into a [BsonValue]. Only primitives, Realm, Bson, lists and maps types
```

Comment:
What about RealmSet, it could be turned into a list for transport?

Reply:
Added support for collections